File: packages/test-db-mcp/src/store.mjs
// @openwork/test-db-mcp 共享的 JSON 测试结果存储（与 test-runner --record 同格式）。
import { mkdir, readFile, writeFile, rename } from "node:fs/promises";
import { existsSync } from "node:fs";
import path from "node:path";
import os from "node:os";
import { randomUUID } from "node:crypto";

export function defaultDbPath(override) {
  if (override) return path.resolve(override);
  if (process.env.OPENWORK_TEST_DB) return path.resolve(process.env.OPENWORK_TEST_DB);
  const home = os.homedir();
  return process.platform === "win32"
    ? path.join(process.env.APPDATA ?? path.join(home, "AppData", "Roaming"), "openwork", "test-results.json")
    : path.join(home, ".openwork", "test-results.json");
}

export class TestDb {
  /** @param {string} file */
  constructor(file) {
    this.file = file;
    /** @type {{runs:any[]}} */
    this.data = { runs: [] };
    this.loaded = false;
  }

  async load() {
    if (this.loaded) return;
    if (existsSync(this.file)) {
      try {
        this.data = JSON.parse(await readFile(this.file, "utf8"));
      } catch {
        this.data = { runs: [] };
      }
    }
    if (!this.data.runs) this.data.runs = [];
    this.loaded = true;
  }

  async flush() {
    await mkdir(path.dirname(this.file), { recursive: true });
    const tmp = `${this.file}.${randomUUID()}.tmp`;
    await writeFile(tmp, JSON.stringify(this.data, null, 2), "utf8");
    await rename(tmp, this.file);
  }

  /** @param {any} run */
  async recordRun(run) {
    await this.load();
    const row = {
      id: run.id ?? `run-${Date.now()}`,
      framework: run.framework ?? "unknown",
      path: run.path ?? ".",
      startedAt: run.startedAt ?? new Date().toISOString(),
      finishedAt: run.finishedAt ?? new Date().toISOString(),
      passed: run.passed ?? 0,
      failed: run.failed ?? 0,
      skipped: run.skipped ?? 0,
      cases: run.cases ?? [],
      coverage: run.coverage ?? null,
    };
    if (run.trigger) row.trigger = run.trigger;
    if (run.taskId) row.taskId = run.taskId;
    if (run.sandbox_id) row.sandbox_id = run.sandbox_id;
    if (run.title) row.title = run.title;
    this.data.runs.push(row);
    await this.flush();
    return row;
  }

  /** @param {{sinceHours?:number, limit?:number}} [opts] */
  async listFailures(opts = {}) {
    await this.load();
    const sinceMs = (opts.sinceHours ?? 24) * 3600 * 1000;
    const cutoff = Date.now() - sinceMs;
    const limit = opts.limit ?? 50;
    /** @type {any[]} */
    const out = [];
    for (let i = this.data.runs.length - 1; i >= 0 && out.length < limit; i--) {
      const r = this.data.runs[i];
      const t = Date.parse(r.startedAt);
      if (Number.isNaN(t) || t < cutoff) continue;
      if (r.failed > 0) {
        out.push({
          ...r,
          cases: (r.cases ?? []).filter((c) => c.status === "failed"),
        });
      }
    }
    return out;
  }

  /** @param {number} [days] */
  async getTrend(days = 7) {
    await this.load();
    const cutoff = Date.now() - days * 24 * 3600 * 1000;
    /** @type {Record<string,{passed:number,failed:number,runs:number}>} */
    const buckets = {};
    for (const r of this.data.runs) {
      const t = Date.parse(r.startedAt);
      if (Number.isNaN(t) || t < cutoff) continue;
      const day = r.startedAt.slice(0, 10);
      if (!buckets[day]) buckets[day] = { passed: 0, failed: 0, runs: 0 };
      buckets[day].passed += r.passed ?? 0;
      buckets[day].failed += r.failed ?? 0;
      buckets[day].runs += 1;
    }
    return Object.entries(buckets)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([day, v]) => ({ day, ...v }));
  }

  async listRuns(limit = 20) {
    await this.load();
    return this.data.runs.slice(-limit).reverse();
  }
}

File: packages/test-runner/internal/store/store.go
// Package store 提供测试结果 JSON 持久化（与 @openwork/test-db-mcp 共用文件格式）。
package store

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

// TestCase 是一条用例结果。
type TestCase struct {
	Name     string `json:"name"`
	Status   string `json:"status"` // passed|failed|skipped
	Duration float64 `json:"durationMs,omitempty"`
	Message  string `json:"message,omitempty"`
	File     string `json:"file,omitempty"`
}

// RunRecord 是一次测试运行记录。
type RunRecord struct {
	ID        string     `json:"id"`
	Framework string     `json:"framework"`
	Path      string     `json:"path"`
	StartedAt string     `json:"startedAt"`
	FinishedAt string    `json:"finishedAt"`
	Passed    int        `json:"passed"`
	Failed    int        `json:"failed"`
	Skipped   int        `json:"skipped"`
	Cases     []TestCase `json:"cases"`
	Coverage  *Coverage  `json:"coverage,omitempty"`
}

// Coverage 是覆盖率摘要。
type Coverage struct {
	Lines   float64 `json:"linesPercent,omitempty"`
	Branches float64 `json:"branchesPercent,omitempty"`
	Raw     any     `json:"raw,omitempty"`
}

// DB 是 JSON 文件后端。
type DB struct {
	path string
}

// File 默认测试结果库路径。
func DefaultPath() string {
	if p := os.Getenv("OPENWORK_TEST_DB"); p != "" {
		return p
	}
	home, _ := os.UserHomeDir()
	if os.Getenv("APPDATA") != "" || filepath.Separator == '\\' {
		base := os.Getenv("APPDATA")
		if base == "" {
			base = filepath.Join(home, "AppData", "Roaming")
		}
		return filepath.Join(base, "openwork", "test-results.json")
	}
	return filepath.Join(home, ".openwork", "test-results.json")
}

// Open 打开或创建 DB。
func Open(path string) (*DB, error) {
	if path == "" {
		path = DefaultPath()
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return nil, err
	}
	return &DB{path: path}, nil
}

type fileData struct {
	Runs []RunRecord `json:"runs"`
}

func (d *DB) load() (fileData, error) {
	var data fileData
	b, err := os.ReadFile(d.path)
	if err != nil {
		if os.IsNotExist(err) {
			return data, nil
		}
		return data, err
	}
	if err := json.Unmarshal(b, &data); err != nil {
		return data, err
	}
	return data, nil
}

func (d *DB) save(data fileData) error {
	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(d.path, b, 0o644)
}

// AppendRun 追加一次运行记录。
func (d *DB) AppendRun(r RunRecord) error {
	data, err := d.load()
	if err != nil {
		return err
	}
	data.Runs = append(data.Runs, r)
	return d.save(data)
}

// ListFailures 返回 since 之后的失败用例。
func (d *DB) ListFailures(since time.Duration) ([]RunRecord, error) {
	data, err := d.load()
	if err != nil {
		return nil, err
	}
	cutoff := time.Now().Add(-since)
	var out []RunRecord
	for i := len(data.Runs) - 1; i >= 0; i-- {
		r := data.Runs[i]
		t, err := time.Parse(time.RFC3339, r.StartedAt)
		if err != nil || t.Before(cutoff) {
			continue
		}
		if r.Failed > 0 {
			failed := r
			failed.Cases = nil
			for _, c := range r.Cases {
				if c.Status == "failed" {
					failed.Cases = append(failed.Cases, c)
				}
			}
			out = append(out, failed)
		}
	}
	return out, nil
}

// Path 返回数据库文件路径。
func (d *DB) Path() string { return d.path }
